### PR TITLE
Fixes #52375. Removes duplicate encoding aliases returned by QgsVectorrDataProvider::availableEncodings()

### LINKED ISSUE #52375: Layer setting - Datasource encoding - UTF-8 twice
### What is the bug or the crash?

In the Datasouce encoding value list I get displayed the value "UTF-8" twice.
What is the difference between the two? Is it a duplicate or is the value e.g. UTF-8 with and without BOM?

qgis3 28 4_data_source_encoding_utf-8_twice

### Steps to reproduce the issue

Select Layer properties
Source - Data source encoding
There is now two times the value UTF-8

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /><style type="text/css">
p, li { white-space: pre-wrap; }
</style></head><body>

QGIS version | 3.28.4-Firenze | QGIS code revision | fd0fb72ab3
-- | -- | -- | --
Qt version | 5.15.3
Python version | 3.9.5
GDAL/OGR version | 3.6.2
PROJ version | 9.1.1
EPSG Registry database version | v10.076 (2022-08-31)
GEOS version | 3.11.1-CAPI-1.17.1
SQLite version | 3.39.4
PDAL version | 2.4.3
PostgreSQL client version | unknown
SpatiaLite version | 5.0.1
QWT version | 6.1.6
QScintilla2 version | 2.13.1
OS version | Windows 10 Version 2009
  |   |   |  
Active Python plugins
m150xml | 2.3.3
pg_history_viewer | 2.2
qgepplugin | dev
QgisModelBaker | v6.5.2
wincan2qgep | 1.1.0
xtflog_checker | v.1.0.2
db_manager | 0.1.20
grassprovider | 2.12.99
MetaSearch | 0.3.6
processing | 2.12.99
sagaprovider | 2.12.99

</body></html>QGIS version
3.28.4-Firenze
QGIS code revision
[fd0fb72ab3](https://github.com/qgis/QGIS/commit/fd0fb72ab3)
Qt version
5.15.3
Python version
3.9.5
GDAL/OGR version
3.6.2
PROJ version
9.1.1
EPSG Registry database version
v10.076 (2022-08-31)
GEOS version
3.11.1-CAPI-1.17.1
SQLite version
3.39.4
PDAL version
2.4.3
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.6
QScintilla2 version
2.13.1
OS version
Windows 10 Version 2009

Active Python plugins
m150xml
2.3.3
pg_history_viewer
2.2
qgepplugin
dev
QgisModelBaker
v6.5.2
wincan2qgep
1.1.0
xtflog_checker
v.1.0.2
db_manager
0.1.20
grassprovider
2.12.99
MetaSearch
0.3.6
processing
2.12.99
sagaprovider
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by TurboGraphxBeige**
Fixes  #52375
------

Why:
------
QgsVectorDataProvider::availableEncodings() is used to populate the Data source encoding combo box in the layer properties dialog. This function uses QT's QTextCodec::availableCodecs() to get the possible encoding aliases, which returns several entries (aliases) for a single encoding.

While it is useful to get aliases for a single encoding, it may be confusing to have exactly the same alias twice or more. For the issue this PR solves, there was "UTF-8" and "UTF-8", which is exactly the same string. 

What:
-------
This PR simply adds to QgsVectorDataProvider::availableEncodings() the std::unique() function from the C++ standard library to remove the identical strings after sorting them.

### CODE REVIEW

**Comment by agiudiceandrea** [comment]
@TurboGraphxBeige, it seems to me that the `_removeDuplicateEncodings` function is not needed if it behaves as the standard binary predicate `==` of the `std::unique` function. Moreover I think an `std::erase `operatition is needed after a `std::unique` operation.

So I would suggest to modify the code in the following ways:

1) to remove the duplicates case-sensitive: just remove the `_removeDuplicateEncodings` function and rewrite last part of the code as:

```cpp
    // Do case-insensitive sorting of encodings and remove duplicates
    std::sort( sEncodings.begin(), sEncodings.end(), _compareEncodings );
    sEncodings.erase( std::unique( sEncodings.begin(), sEncodings.end() ), sEncodings.end() );
```
or as:
```cpp
    // Do case-insensitive sorting of encodings and remove duplicates
    std::sort( sEncodings.begin(), sEncodings.end(), _compareEncodings );
    sEncodings.removeDuplicates();
```


2) to remove the duplicates case-insensitive:

```
static bool _removeDuplicateEncodings( const QString &s1, const QString &s2 )
{
  return s1.toLower() == s2.toLower();
}

.....

    // Do case-insensitive sorting of encodings then remove duplicates
    std::sort( sEncodings.begin(), sEncodings.end(), _compareEncodings );
    const auto last = std::unique( sEncodings.begin(), sEncodings.end(), _removeDuplicateEncodings );
    sEncodings.erase( last, sEncodings.end() );
```

**Comment by TurboGraphxBeige** [comment]
Yes you're totally right, I didn't realize the exact behavior of std::unique(). Should have tested more, sorry for the rookie mistake.

**Comment by nyalldawson** [comment]
Can you remove the srs db file from this pr?

**Comment by TurboGraphxBeige** [comment]
> Can you remove the srs db file from this pr?

Done

File: src/core/vector/qgsvectordataprovider.cpp [hunk 1]
```diff
@@ -743,6 +743,11 @@ static bool _compareEncodings( const QString &s1, const QString &s2 )
   return s1.toLower() < s2.toLower();
 }
 
+static bool _removeDuplicateEncodings( const QString &s1, const QString &s2 )
+{
+  return ( s1.toLower() == s2.toLower() );
```
Comment (nyalldawson) [inline src/core/vector/qgsvectordataprovider.cpp:748]:
```suggestion
  return ( s1.compare( s2, Qt::CaseInsensitive ) == 0 );
```

File: src/core/vector/qgsvectordataprovider.cpp [hunk 2]
```diff
@@ -743,6 +743,11 @@ static bool _compareEncodings( const QString &s1, const QString &s2 )
   return s1.toLower() < s2.toLower();
 }
 
+static bool _removeDuplicateEncodings( const QString &s1, const QString &s2 )
+{
+  return ( s1.compare( s2, Qt::CaseInsensitive ) == 0 );
```
Comment (nyalldawson) [inline src/core/vector/qgsvectordataprovider.cpp:748]:
```suggestion
  return s1.compare( s2, Qt::CaseInsensitive ) == 0;
```

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**